Device code flow

skabou opened the pull request:
#DOCS-CODE-031 Initial check-in of Device Code Flow in PowerShell

skabou committed:
Create app.ps1

Initial check-in

skabou committed:
Update app.ps1

skabou committed:
Create README.md

skabou committed:
Update README.md

skabou committed:
Update app.ps1

skabou committed:
Update app.ps1

skabou committed:
Update app.ps1

skabou committed:
Update app.ps1

skabou committed:
Update app.ps1

skabou committed:
Update app.ps1

ckittel commented on device-code-flow/app.ps1:
```diff
@@ -0,0 +1,66 @@
+$TenantId = ""
+
+# Configure $DeviceCodeRequestParams for the desired $TenantId and $ClientId
+$DeviceCodeRequestParams = @{
+    Method = "POST"
+    Uri    = "https://login.microsoftonline.com/$TenantId/oauth2/devicecode"
+    Body   = @{
+        resource  = "https://graph.microsoft.com"
+        client_id = $ClientId
+    }
+}
```
Typically, we'd expect users to use MSAL.NET from PowerShell.  One of PowerShell's key features is to use .NET assemblies as part of its execution, so that you don't need to replicate logic within PowerShell that already has a .NET SDK.  Even our unsupported MSAL.PS module uses MSAL.NET under the covers.  There is just too much error handling, token caching, etc that is being left on the table when trying to talk to Azure AD without the msal.net library.

Any reason to avoid it that you came across?  Even the Azure SDK for powershell uses `Azure.Identity` which itself uses MSAL.NET.

skabou replied:
No reason.  Will now look into integrating MSAL.NET

ckittel replied:
Resolving this comment since we have a direction here.

ckittel reviewed (changes requested):
Just wanted to get a clarification before I went further in the review, so pausing here until I know if this is the expected path.

skabou committed:
Work In Progress

Updated the code to use MSAL.PS  (just as an example)
Trying to get it to work without MSAL.PS

skabou committed:
Update app.ps1

Update to remove MSAL.PS and include a C# helper function

skabou committed:
Update README.md

skabou committed:
Update README.md

skabou committed:
Update README.md

skabou committed:
Update README.md

skabou committed:
Update app.ps1

skabou committed:
Create DeviceCodeHelper.cs

skabou committed:
Update app.ps1

Moving C# code to its own file

skabou committed:
Update README.md

skabou committed:
Update app.ps1

skabou committed:
Update app.ps1

skabou committed:
Update app.ps1

ferantivero commented on device-code-flow/README.md:
```diff
@@ -0,0 +1,128 @@
+
+- Azure Active Directory (Azure AD) tenant and the permissions or role required for managing app registrations in the tenant.
+- PowerShell
+- MSAL.NET
```
Im personally not clear whether a package/module is a pre-req. I've seen some readme template that has the packages installation as part of the `Run the app` section.

skabou replied:
Thank you for this.  I was not sure whether to include the step for installing MSAL.NET.  I can add it, it's a one line operation.

ferantivero commented on device-code-flow/README.md:
```diff
@@ -0,0 +1,128 @@
+## Prerequisites
+
+- Azure Active Directory (Azure AD) tenant and the permissions or role required for managing app registrations in the tenant.
+- PowerShell
```
have we considered targeting PowerShell core?

skabou replied:
Not that I'm aware of 🤔

ferantivero replied:
That will be our users the opportunity to run this on Linux or macOS

skabou replied:
@ferantivero Updated the code to work in PowerShell 7 per the original ask 😄

ferantivero commented on device-code-flow/DeviceCodeHelper.cs:
```diff
@@ -0,0 +1,26 @@
+            // * The timeout specified by the server for the lifetime of this code (typically ~15 minutes) has been reached
+            // * The developing application calls the Cancel() method on a CancellationToken sent into the method.
+            //   If this occurs, an OperationCanceledException will be thrown (see catch below for more details).
+            Console.WriteLine(deviceCodeResult.Message);
```
it seems this class has a newer version but more importantly it is already part of [the module itself](https://www.powershellgallery.com/packages/MSAL.PS/4.37.0.0/Content/internal%5CDeviceCodeHelper.cs). I was wondering whether it is really required.

skabou replied:
So the link you provided is for MSAL.PS which is a project not supported by Microsoft.
While I was able to get a tutorial working with MSAL.PS I was directed to not use it.

For this tutorial we only want to use Microsoft Authentication Library for .NET (MSAL.NET)

ferantivero replied:
Oh I see, if that is not supported by MS sounds like a good idea not to use it for now. 

Please lets consider this one as solved.

ferantivero reviewed (changes requested):
nicely done @skabou couple of comments below to be reviewed.

skabou committed:
Update README.md

Removed "MSAL.NET" as a prerequisite, and instead added it as a step, instructing how to install it.

skabou commented:
@ckittel @ferantivero  I just wanted to point out...  Right now I have a separate file for DeviceCodeHelper.cs - I also have another version that embeds the C# code with the powershell script (instead of using DeviceCodeHelper.cs).  I was't sure which way was preferred for this tutorial.

ferantivero commented:
> @ckittel @ferantivero I just wanted to point out... Right now I have a separate file for DeviceCodeHelper.cs - I also have another version that embeds the C# code with the powershell script (instead of using DeviceCodeHelper.cs). I was't sure which way was preferred for this tutorial.

yeah, I'd rather embed since originally I was looking for something that could enable in-lining.

skabou committed:
Delete DeviceCodeHelper.cs

skabou committed:
Update app.ps1

Put the C# code back in-line and removed DeviceCodeHelper.cs

skabou commented:
> > @ckittel @ferantivero I just wanted to point out... Right now I have a separate file for DeviceCodeHelper.cs - I also have another version that embeds the C# code with the powershell script (instead of using DeviceCodeHelper.cs). I was't sure which way was preferred for this tutorial.
> 
> yeah, I'd rather embed since originally I was looking for something that could enable in-lining.

Updated - removed DeviceCodeHelper.cs and used in-line code instead

skabou committed:
Update README.md

updates to 'About the code'

skabou committed:
Update app.ps1

Updated for compatibility with PowerShell 7

skabou committed:
Update README.md

skabou commented:
I believe this is ready for review again.  Needed to make a few adjustments to get it to work with Powershell 7 @ckittel

ckittel commented on device-code-flow/README.md:
```diff
@@ -0,0 +1,135 @@
+## Prerequisites
+
+- Azure Active Directory (Azure AD) tenant and the permissions or role required for managing app registrations in the tenant.
+- PowerShell 7+
```
Because you're using .NET, do you need the .NET 6(?) runtime as well?  Or does this come automatically when you do the `Install-Package` call?

skabou replied:
It appears the .NET runtime is not required

skabou replied:
I did not need to install the .NET SDK nor .NET runtime for this to work.

ckittel replied:
Coo.

ckittel commented on device-code-flow/app.ps1:
```diff
@@ -0,0 +1,63 @@
+$TenantId = ""
+
+# Scope permission for Graph
+[string[]] $Scope = "User.Read"
```
Any chance you can inline this on line 48?  We've inlined this in most of the other ones so that it's easy to see (when taken as a local snippet), what scope is being asked for.

skabou replied:
Yep - fix coming

skabou replied:
updated

ckittel replied:
looks good!

ckittel commented on device-code-flow/app.ps1:
```diff
@@ -0,0 +1,63 @@
+    Method  = "GET"
+    Uri     = "https://graph.microsoft.com/v1.0/me"
+    Headers = @{
+        "Authorization" = "Bearer " + $TokenRequest.AccessToken
```
string interpolation works in PS, yeah?

```suggestion
        "Authorization" = "Bearer $TokenRequest.AccessToken"
```

skabou replied:
I had to look it up but apparently this is the way:

"Authorization" = "Bearer $($TokenRequest.AccessToken)"

skabou replied:
But the other mechanism overrides this

ckittel replied:
Cool, you were able to get the other way to work.  Nice.

ckittel commented on device-code-flow/app.ps1:
```diff
@@ -0,0 +1,63 @@
+}
+
+# Send a request to the Graph API with the token to retrieve the values from /me
+$GraphRequest = Invoke-RestMethod @GraphRequestParams
```
This would be "$GraphResponse," right?

skabou replied:
updated

ckittel commented on device-code-flow/app.ps1:
```diff
@@ -0,0 +1,63 @@
+$publicClient = [Microsoft.Identity.Client.PublicClientApplicationBuilder]::Create($ClientId).WithTenantId($TenantId)
+
+# Acquire an AccessToken for the specified scope
+$TokenRequest = $publicClient.Build().AcquireTokenWithDeviceCode($Scope, [DeviceCodeHelper]::GetDeviceCodeResultCallback()).ExecuteAsync().Result
```
This would be "$TokenResponse," right?

skabou replied:
updated

ckittel commented on device-code-flow/app.ps1:
```diff
@@ -0,0 +1,63 @@
+$GraphRequestParams = @{
+    Method  = "GET"
+    Uri     = "https://graph.microsoft.com/v1.0/me"
+    Headers = @{
+        "Authorization" = "Bearer " + $TokenRequest.AccessToken
+    }
```
I think the params have a mechanism for this without doing it manually, it's something like the following, but I don't remember specifically. 

```suggestion
    Authentication = "Bearer"
    Token          = (ConvertTo-SecureString $TokenRequest.AccessToken)
```

skabou replied:
implemented

ckittel replied:
Cool, you were able to get this way to work.  I was a bit off on my guess, but looks like you got it figured out. Nice.

ckittel reviewed (changes requested):
w00t.  Here are a couple items I noticed that you might want to consider.

skabou committed:
Update app.ps1

$TokenRequest -> $TokenResponse
$GraphRequest -> $GraphResponse
Replaced "Authorization" with "Authentication"+"Token"
Moved scope to be in-line

ckittel reviewed (approved):
Heck yeah!  🚀

mmacy commented on device-code-flow/README.md:
```diff
@@ -0,0 +1,135 @@
+| **Platform type**                 | _None_                                                                       | No redirect URI required; don't select a platform.                                                 |
+| **Allow public client flows**     | **Yes**                                                                      | Required value for this sample.                                                                    |
+
+> :information_source: **Bold text** in the tables above matches (or is similar to) a UI element in the Azure portal, while `code formatting` indicates a value you enter into a text box in the Azure portal.
```
```suggestion
> :information_source: **Bold text** in the table above matches (or is similar to) a UI element in the Azure portal, while `code formatting` indicates a value you enter into a text box in the Azure portal.
```

mmacy commented on device-code-flow/README.md:
```diff
@@ -0,0 +1,135 @@
+
+### 3. Install package(s)
+
+To install MSAL.NET libraries in PowerShell:
```
```suggestion
Install the MSAL.NET libraries in PowerShell:
```

mmacy reviewed (approved):
Couple super minor nits, otherwise, LTGM. Thanks, @skabou! :shipit: 

> EDIT:  Just noticed the README was missing the sample ID comment:
> https://github.com/Azure-Samples/ms-identity-docs-code-powershell/pull/1/files#r827601305

mmacy commented on device-code-flow/README.md:
```diff
@@ -0,0 +1,135 @@
+urlFragment: ms-identity-docs-code-app-device-code-powershell
+---
+-->
+
```
```suggestion
<!-- SAMPLE ID: DOCS-CODE-031 -->
```

skabou committed:
Update device-code-flow/README.md

Co-authored-by: Marsh Macy <mmacy@users.noreply.github.com>

skabou committed:
Update device-code-flow/README.md

Co-authored-by: Marsh Macy <mmacy@users.noreply.github.com>

skabou committed:
Update device-code-flow/README.md

Co-authored-by: Marsh Macy <mmacy@users.noreply.github.com>
